Automattic/jetpack
# Connection: fix user connection retry flow

Related to https://github.com/Automattic/wp-calypso/issues/55969

If the first attempt to authorize the user in Calypso fails, they get redirected to the Jetpack site for a new authorization URL.
The handler is located in Jetpack, so other plugins (e.g. Backup) cannot forward the user to reattempt the authorization if Jetpack-the-plugin is not installed.

The PR moves the handler into Connection package, allowing the code to perform on any Connection package consumer.

#### Changes proposed in this Pull Request:
* Move the `connect_url_redirect` action handling into Connection package's webhooks

#### Jetpack product discussion
p9dueE-5cY-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
First, reproduce the issue before checking out the branch:
1. Disconnect and deactivate Jetpack, log out of WP.com
2. Install Jetpack Backup, initiate the connection.
3. During the connection process, login, then click "Approve" button. You should see an error for a few seconds, then you'll get redirected.
4. You will get redirected to an empty page with the URL `wp-admin/admin.php?page=jetpack&connect_url_redirect=true&...`
5. Checkout the branch and repeat the steps 1-3. Instead of an empty wp-admin page, you should get redirected back to Calypso. Click the "Approve" button and confirm that you're connected and redirected back to the `?page=jetpack-backup`.

#### Connection flow testing scenarios
_These are mostly for myself, to make sure I've covered all the connection flows. You may choose a few to test if you've got time._

|| Scenario | Plugin | Active Jetpack | Disabled Jetpack |
| - | ----------- | ----------- | ----------- | ----------- |
| ✅ | Logged into WP.com | Jetpack | x | -- | 
| ✅ | Need to log into WP.com | Jetpack | x | -- | 
| ✅ | Need to register on WP.com | Jetpack | x | -- | 
| ✅ | Connecting the user in another window before clicking "Approve" | Jetpack | x | -- | 
| ✅ | After connecting the user, removing the `calypso_env` parameter from the `connect_url_redirect` URL | Jetpack | x | -- |
| ✅ | Logged into WP.com | Backup | x | x | 
| ✅ | Need to log into WP.com | Backup | x | x | 
| ✅ | Need to register on WP.com | Backup | x | x | 
| ✅ | Connecting the user in another window before clicking "Approve" | Backup | x | x | 
| ✅ | After connecting the user, removing the `calypso_env` parameter from the `connect_url_redirect` URL | Backup | x | x |
| ✅ | Logged into WP.com | MyJetpack | x | x | 
| ✅ | Need to log into WP.com | MyJetpack | x | x | 
| ✅ | Need to register on WP.com | MyJetpack | x | x | 
| ✅ | Connecting the user in another window before clicking "Approve" | MyJetpack | x | x | 
| ✅ | After connecting the user, removing the `calypso_env` parameter from the `connect_url_redirect` URL | MyJetpack | x | x |
